### PR TITLE
switch code quality workflow to dev target and smoketest

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
This PR does two things:
- switches our code quality workflow back to dev target to avoid forcing install all for linting to pass
- adds a smoketest workflow that just imports the top level composer modules to avoid accidentally missing conditional imports

# What issue(s) does this change relate to?
Closes [CO-1866](https://mosaicml.atlassian.net/browse/CO-1866)

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1866]: https://mosaicml.atlassian.net/browse/CO-1866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by dakinggg** [comment]
Before the import fix from hanlin's pr, running the smoketest locally
```
(composer-dev-3.10) danielking@MML-1B940F4333E2 composer % pytest tests/test_smoketest.py 
ImportError while loading conftest '/Users/danielking/github/composer/tests/conftest.py'.
tests/conftest.py:9: in <module>
    from composer.utils import reproducibility
composer/__init__.py:10: in <module>
    from composer.trainer import Trainer
composer/trainer/__init__.py:6: in <module>
    from composer.trainer.trainer import Trainer
composer/trainer/trainer.py:33: in <module>
    from composer.callbacks import CheckpointSaver, OptimizerMonitor
composer/callbacks/__init__.py:12: in <module>
    from composer.callbacks.health_checker import HealthChecker
composer/callbacks/health_checker.py:18: in <module>
    import slack_sdk
E   ModuleNotFoundError: No module named 'slack_sdk'
```

after
```
(composer-dev-3.10) danielking@MML-1B940F4333E2 composer % pytest tests/test_smoketest.py
=================================================================================== test session starts ===================================================================================
platform darwin -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
rootdir: /Users/danielking/github/composer, configfile: pyproject.toml
plugins: pytest_codeblocks-0.16.1, anyio-3.6.2, httpserver-1.0.6
collected 1 item                                                                                                                                                                          

tests/test_smoketest.py .                                                                                                                                                           [100%]

==================================================================================== 1 passed in 0.02s ====================================================================================
(composer-dev-3.10) danielking@MML-1B940F4333E2 composer % 
```

**Review by bandish-shah — APPROVED** [review]
LGTM, thanks for adding this simple but likely very useful sanity check.

File: tests/test_smoketest.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+from composer import (algorithms, callbacks, core, datasets, devices, functional, loggers, loss, metrics, models, optim,
+                      profiler, trainer, utils)
+
+
+# this very simple test is just to use the above imports, which check and make sure we can import all the top-level
```
Comment (mvpatel2000) [inline tests/test_smoketest.py:8]:
```suggestion
# This very simple test is just to use the above imports, which check and make sure we can import all the top-level
```

File: .github/workflows/smoketest.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+jobs:
+  smoketest:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 10
+    strategy:
+      matrix:
+        python_version:
+          - "3.8"
+          - "3.9"
+          - "3.10"
```
Comment (mvpatel2000) [inline .github/workflows/smoketest.yaml:31]: do we need to run on all? I suppose it doesn't hurt...

Comment (dakinggg) [reply .github/workflows/smoketest.yaml:31]: yeah, its super fast. seems fine

**Pull request merged**